source/extensions/filters/http/cache/cache_filter.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "envoy/config/filter/http/cache/v2alpha/cache.pb.h"
+
+#include "extensions/filters/http/cache/http_cache.h"
+#include "extensions/filters/http/common/pass_through_filter.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+
+/**
+ * A filter that caches responses and attempts to satisfy requests from cache.
+ * Must be owned by a shared_ptr.
+ */
+class CacheFilter;
+using CacheFilterSharedPtr = std::shared_ptr<CacheFilter>;
+class CacheFilter : public Http::PassThroughFilter,
+                    public std::enable_shared_from_this<CacheFilter> {
```

Comment:
TIL: this exists!  I did not know how to do this using shared_ptr.

Do you know what overhead this adds? I doesn't matter for this class as there won't be a lot of instances but I'm curious in general.

Reply:
The libstdcxx implementation of std::enable_shared_from_this increases the class's size by two pointers, and calling shared_from_this() will do an atomic op or two. There shouldn't be any overhead beyond that.

The biggest downside is the requirement that CacheFilter now has to always be allocated with std::make_shared.

Reply:
Actually 2 things:

1. I'm not sure if this is useful but an alternate strategy that exists now (but didn't previously) is in include/envoy/stats/refcount_ptr.h. This is a bit slimmer than using enable_shared_from_this and shared_ptr. Probably not that important but I thought it mentioned it exists.

2. It's worth commenting why this is needed. E.g. are we keeping cache objects alive until async operations pending on them are complete?

Reply:
CacheFilterSharedPtr must be convertible to Envoy::Http::StreamFilterSharedPtr to be passed to FilterChainFactoryCallbacks::addStreamFilter, so RefcountPtr isn't an option.

Reply:
Comment added.